src/serac/physics/utilities/functional/integral.hpp

```diff
@@ -569,11 +577,7 @@ class Integral {
     using u_du_t          = typename detail::lambda_argument<trial_space, geometry_dim, spatial_dim>::type;
     using derivative_type = decltype(get_gradient(qf(x_t{}, make_dual(u_du_t{}))));
 
-    auto num_quadrature_points = static_cast<uint32_t>(X.Size() / spatial_dim);
-    qf_derivatives_.resize(sizeof(derivative_type) * num_quadrature_points);
-
-    // FIXME: Strict aliasing rule?
-    auto qf_derivatives_ptr = reinterpret_cast<derivative_type*>(qf_derivatives_.data());
+    std::shared_ptr<derivative_type[]> qf_derivatives(new derivative_type[num_quadrature_points]);
```

Comment:
Why does this need to be a `shared_ptr`?

Reply:
I think that's what this comment refers to:

https://github.com/LLNL/serac/pull/418#discussion_r628635383

Instead of performing a (dubious) reinterpret_cast, this approach avoids type punning entirely.

Reply:
Though I'm concerned the lifetime of `qf_derivatives` ends at the end of the constructor - perhaps a `shared_ptr` should be the type of the argument to the kernels so the reference count is incremented appropriately?

Reply:
`qf_derivatives` can't be a member variable (`derivative_type` is specific to the lambda function, so it can't be known a priori), so there's no way to `delete` it in the destructor.

By making it a shared pointer, its lifetime should essentially match the lifetime of the `Integral` that created it:
e.g. 
1. Integral constructor: creates `qf_derivatives` and the two lambda functions capture it, incrementing the reference count.
2. Integral destructor: `std::function`s that capture `qf_derivatives` get deallocated, and the reference count of `qf_derivatives` goes to zero again, triggering its cleanup.

The previous implementation involved essentially making a member variable `void * qf_derivatives` and aliasing that pointer in a way that isn't really type safe. 

If there's a better way to do it, I'm interested! This was just one implementation that came to mind.

Reply:
I'd add a comment reflecting this. I can't think of a better way to do this at the moment. Thank you for explaining!